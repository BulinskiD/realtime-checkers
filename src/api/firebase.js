import app from "firebase/app";
import "firebase/storage";
import "firebase/auth";
import "firebase/firestore";

var firebaseConfig = {
  apiKey: "AIzaSyD3uWma_iug5o_0w6EmchIXp4qd054HqcE",
  authDomain: "realtime-checkers.firebaseapp.com",
  databaseURL: "https://realtime-checkers.firebaseio.com",
  projectId: "realtime-checkers",
  storageBucket: "realtime-checkers.appspot.com",
  messagingSenderId: "100295366319",
  appId: "1:100295366319:web:ec1c182d83b660e8"
};

// Initialize Firebase
app.initializeApp(firebaseConfig);

export const auth = app.auth();
export const firestore = app.firestore();
