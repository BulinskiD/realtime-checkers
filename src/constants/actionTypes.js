export const TEST_ACTION = 'TEST_ACTION';
export const LOGIN_SUCCESS = 'LOGIN_SUCCESS';
