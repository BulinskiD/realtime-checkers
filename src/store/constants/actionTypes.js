export const START_LOADING = 'START_LOADING';
export const STOP_LOADING = 'STOP_LOADING';
export const LOGIN_SUCCESS = 'LOGIN_SUCCESS';
export const LOGIN_FAILED = 'LOGIN_FAILED';
export const LOGOUT_SUCCESS = 'LOGOUT_SUCCESS';
export const LOGOUT_FAILED = 'LOGOUT_FAILED';
