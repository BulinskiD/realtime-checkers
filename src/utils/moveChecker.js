export default (checkersPosition, from, to) => {

    let row = from.color === 'black' ? from.row - 1 : from.row + 1;


    if((from.col - to.col) === 2) {
        checkersPosition = checkersPosition.filter(item => {
            if((from.col - 1 === item.col) && (row === item.row)) {
                return item.color === from.color;
            }
            return true;
        });
    }

    if((from.col - to.col) === -2) {
        checkersPosition = checkersPosition.filter(item => {
            if((from.col + 1 === item.col) && (row === item.row)) {
                return item.color === from.color;
            }
            return true;
        });
    }



    checkersPosition[checkersPosition.indexOf(from)].col = to.col;
    checkersPosition[checkersPosition.indexOf(from)].row = to.row;
    checkersPosition[checkersPosition.indexOf(from)].selected = false;

    return checkersPosition;
}