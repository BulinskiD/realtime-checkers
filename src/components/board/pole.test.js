import React from "react";
import { shallow } from "enzyme";
import moveChecker from "../../utils/moveChecker";
import { firestore } from "../../api/firebase";
import ConnectedPole, { Pole } from "./pole";
import configureMockStore from "redux-mock-store";
import {
  getPole,
  checkNextStatus,
  checkIfWinner
} from "../../utils/utilFunctions";
import handleError from "../../utils/handleError";
import { BLACK_WINNER } from "../../store/constants/actionTypes";

jest.mock("../../utils/handleError");
jest.mock("../../utils/utilFunctions");
jest.mock("../../api/firebase");
jest.mock("../../utils/moveChecker");

Date.now = jest.fn();
Date.now.mockReturnValue(123);

const mockedStore = configureMockStore();
checkIfWinner.mockReturnValue(null);
const doc = jest.fn();
const update = jest.fn();
firestore.collection = jest.fn();
firestore.collection.mockReturnValue({ doc });
doc.mockReturnValue({ update });

describe("Pole", () => {
  let props = {};
  beforeEach(() => {
    props = {
      currentGame: {
        id: "2",
        status: "white",
        nextMove: true,
        isActiveTurn: true,
        checkersPosition: []
      },
      pole: { color: "black" },
      active: false,
      col: 1,
      row: 0
    };
  });
  it("should render black checker for given props", () => {
    const component = shallow(<Pole {...props} />);
    expect(component).toMatchSnapshot();
  });

  it("should render active pole for given props", () => {
    props.pole = null;
    props.active = true;
    const component = shallow(<Pole {...props} />);

    expect(component.find("Connect(Checker)").exists()).toBeFalsy();
    expect(component.find(".active").exists()).toBeTruthy();
    expect(component).toMatchSnapshot();
  });

  it("should handle move properly", async () => {
    const selectChecker = jest.fn();
    const setActivePoles = jest.fn();
    const newProps = {
      ...props,
      pole: null,
      active: false,
      selectChecker,
      setActivePoles
    };
    moveChecker.mockReturnValue({
      checkersPosition: [],
      hasNextMove: false,
      selectedChecker: {}
    });
    checkNextStatus.mockReturnValue("black");
    const prom = Promise.resolve("OK");
    update.mockReturnValue(prom);

    const component = shallow(<Pole {...newProps} />);

    component.find(".active").simulate("click");

    await prom;
    expect(update).toBeCalledWith({
      status: "black",
      checkersPosition: [],
      selectedChecker: {},
      nextMove: false,
      from: {},
      updated: 123
    });
  });

  it("should set status BLACK_WINNER for given input", async () => {
    const selectChecker = jest.fn();
    const setActivePoles = jest.fn();
    const props = {
      id: "2",
      status: "white",
      isActiveTurn: true,
      nextMove: true,
      checkersPosition: [],
      pole: null,
      active: true,
      col: 1,
      row: 0,
      selectChecker,
      setActivePoles
    };
    moveChecker.mockReturnValue({
      checkersPosition: [],
      hasNextMove: false,
      selectedChecker: {}
    });
    checkNextStatus.mockReturnValue("black");
    const prom = Promise.resolve("OK");
    update.mockReturnValue(prom);
    checkIfWinner.mockReturnValue(BLACK_WINNER);

    const component = shallow(<Pole {...props} />);

    component.find(".active").simulate("click");

    await prom;
    expect(update).toBeCalledWith({
      checkersPosition: [],
      selectedChecker: {},
      status: BLACK_WINNER,
      nextMove: false,
      from: {},
      updated: 123
    });
  });

  it("should call handle error when request fails", async () => {
    expect.assertions(2);

    const selectChecker = jest.fn();
    const setActivePoles = jest.fn();
    const props = {
      id: "2",
      status: "white",
      isActiveTurn: true,
      nextMove: true,
      checkersPosition: [],
      pole: null,
      active: true,
      col: 1,
      row: 0,
      selectChecker,
      setActivePoles
    };
    moveChecker.mockReturnValue({
      checkersPosition: [],
      hasNextMove: false,
      selectedChecker: {}
    });
    checkNextStatus.mockReturnValue("black");
    const prom = Promise.reject("Error");
    update.mockReturnValue(prom);

    const component = shallow(<Pole {...props} />);

    component.find(".active").simulate("click");
    expect(update).toBeCalledWith({
      status: "black",
      checkersPosition: [],
      selectedChecker: {},
      nextMove: false,
      from: {},
      updated: 123
    });
    try {
      await prom;
    } catch (error) {
      expect(handleError).toBeCalledWith("Error");
    }
  });
});

describe("MapStateToProps in Pole", () => {
  it("should map state to props correctly", () => {
    const storeState = {
      currentGame: {
        status: "white",
        selectedChecker: {},
        activePoles: [{ col: 2, row: 2 }],
        id: "22",
        checkersPosition: [{ col: 2, row: 2, color: "black", selected: false }],
        nextMove: false
      }
    };

    const props = {
      id: "22",
      status: "white",
      nextMove: false,
      checkersPosition: [{ col: 2, row: 2, color: "black", selected: false }],
      pole: { col: 2, row: 2 },
      active: true,
      col: 2,
      row: 2,
      selectedChecker: {}
    };

    getPole.mockReturnValue({ col: 2, row: 2 });
    const store = mockedStore(storeState);
    const component = shallow(<ConnectedPole store={store} col={2} row={2} />);
    expect(component.find("Pole").props()).toEqual(
      expect.objectContaining({ ...props })
    );
  });
});
