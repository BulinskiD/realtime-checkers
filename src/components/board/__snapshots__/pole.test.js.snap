// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pole should render active pole for given props 1`] = `
<styled.div
  col={1}
  row={0}
>
  <styled.div
    className="active"
    onClick={[Function]}
  />
</styled.div>
`;

exports[`Pole should render black checker for given props 1`] = `
<styled.div
  col={1}
  row={0}
>
  <Connect(Checker)
    color="black"
    isActiveTurn={true}
    status="white"
  />
</styled.div>
`;
